make coarsen reductions consistent with reductions on other classes

dcherian opened the pull request:
- [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

This PR uses `inject_reduce_methods` to inject reduction methods into the Coarsen classes. So now we can do `coarsen.count()` and pass `skipna` down to the reduction methods.

max-sixty commented:
Great idea, there's lots of room to centralize some of these!

dcherian commented:
@fujisoup this could use a review if you have time.

fujiisoup commented on xarray/core/variable.py:
```diff
@@ -1843,11 +1843,11 @@ def coarsen(self, windows, func, boundary="exact", side="left"):
             if func is None:
                 raise NameError(f"{name} is not a valid method.")
-        return type(self)(self.dims, func(reshaped, axis=axes), self._attrs)
+        return type(self)(self.dims, func(reshaped, axis=axes, **kwargs), self._attrs)
```
Probably now we can use `_replace` method?

dcherian commented:
> I don't remember why I added a separate injection function...

:+1:

> Can you add some tests for coarsen.count() and skipna in the reduction methods?

There is a `count` test in `test_variable`. I've added `sum(skipna=...)` tests there too now.

keewis commented on doc/whats-new.rst:
```diff
@@ -85,6 +85,9 @@ New Features
   `Guido Imperiale <https://github.com/crusaderky>`_.
+- Added the :py:meth:`count` reduction method to both :py:class:`DatasetCoarsen`
+  and :py:class:`DataArrayCoarsen` objects. (:pull:`3500`)
+  By `Deepak Cherian <https://github.com/dcherian/>`_
```
now that #3516 is merged, this should be
```suggestion
  By `Deepak Cherian <https://github.com/dcherian>`_
```

dcherian replied:
thanks!

dcherian commented:
> Probably I wondered if bottleneck works with multiple reduction axes.

I'll check this before merging.

dcherian commented:
In the case of multiple reduction axes i.e. `isinstance(axis, tuple) is True` we don't use the bottleneck method so things should be fine.

https://github.com/pydata/xarray/blob/ed05f9862622b00f40f7b9b99ccdb0ab3766ff0f/xarray/core/nputils.py#L210-L218
